### ISSUE TITLE
Run an HTML validator as part of the test suite

### ISSUE THREAD

**Issue by aantron** [ocaml/odoc#165]

**Comment by pmetzger** [comment]
Do you have one in particular in mind?

**Comment by aantron** [comment]
No, we need to pick one. Ideally something easy to install on multiple platforms. If not, then it should be easy to install in Ubuntu in Travis.

**Comment by rizo** [comment]
I'd add that it should support the heading-level outline algorithm similar to one in https://validator.w3.org/nu.

**Comment by aantron** [comment]
That makes me wonder if there is a validator we can access using curl/wget.

**Comment by dbuenzli** [comment]
The one @rizo mentioned is available as a cli tool (but you'll need java). Other than that there's at least [`tidy`](http://www.html-tidy.org/). Which I is available in brew as `tidy-html5` and in ubuntu as [`tidy`](https://packages.ubuntu.com/bionic/tidy).

**Comment by ostera** [comment]
+1 for using `tidy` as well for the expected outputs. Will make updating the expected HTML in PRs much clearer as well.

**Comment by pmetzger** [comment]
If the consensus is tidy, what do we need to do in order to make that happen?

**Comment by rizo** [comment]
I think before the integration, the first step would be to run tidy on the existing [HTML expected files](https://github.com/ocaml/odoc/tree/master/test/html/expect/ml) and find all issues with it. Those can be reported as issues and fixed separately. After that we can update the expected files and integrate tidy as part of the validation for actual output.

**Comment by ostera** [comment]
After running:

```sh
tidy -ashtml test/html/expect/*/*.html 2>&1
  | grep line \
  | awk -F':' '{ print $2 }' \
  | sort -u
```

I see the following problems:

  1. `<h2>` anchor "empty-section" already defined
  2. discarding unexpected `</dt>`
  3. missing `<dd>`
  4. missing optional end tag `</dd>`
  5. nested emphasis `<em>`
  6. trimming empty `<code>`

Seems small enough to be addressed in a single PR.

**Comment by ostera** [comment]
This can be closed now.

**Comment by aantron** [comment]
Thanks, @ostera :)

Addressed by @ostera in #185 and #190.
